diff --git a/configure b/configure
index aa9a7da..20069fc 100755
--- a/configure
+++ b/configure
@@ -3704,7 +3704,7 @@ fi
 # check if memfd is supported
 memfd=no
 cat > $TMPC << EOF
-#include <sys/memfd.h>
+#include <sys/mman.h>
 
 int main(void)
 {
diff --git a/util/memfd.c b/util/memfd.c
index 4571d1a..6dd5e72 100644
--- a/util/memfd.c
+++ b/util/memfd.c
@@ -32,7 +32,7 @@
 #include "qemu/memfd.h"
 
 #ifdef CONFIG_MEMFD
-#include <sys/memfd.h>
+#include <sys/mman.h>
 #elif defined CONFIG_LINUX
 #include <sys/syscall.h>
 #include <asm/unistd.h>
