diff --git a/core/swupdate.c b/core/swupdate.c
index e08659e..651b082 100644
--- a/core/swupdate.c
+++ b/core/swupdate.c
@@ -927,5 +927,15 @@ int main(int argc, char **argv)
 	if (!opt_c && !opt_i)
 		pthread_join(network_daemon, NULL);
 
+	/*
+	 * Remove the socket after operation
+	 */
+	if (!opt_c) {
+		if (unlink(SOCKET_PROGRESS_PATH) < 0) {
+			ERROR("Error removing IPC socket %s.", SOCKET_PROGRESS_PATH);
+			exit(1);
+		}
+	}
+
 	return result;
 }
